MapboxMobileEvents/MMEEvent.m

```diff
@@ -3,305 +3,336 @@
+
+@end
+
+#pragma mark -
+
 @implementation MMEEvent
 
-+ (instancetype)eventWithDate:(MMEDate *)eventDate name:(NSString *)eventName attributes:(NSDictionary *)attributes {
-    return [MMEEvent.alloc initWithDate:eventDate name:eventName attributes:attributes];
++ (NSDictionary *)nilAttributes {
+    static NSDictionary *nilAttributes = nil;
+    if (!nilAttributes) {
+        nilAttributes = @{};
+    }
+    return nilAttributes;
 }
 
-+ (instancetype)eventWithName:(NSString *)eventName attributes:(NSDictionary *)attributes {
-    return [MMEEvent eventWithDate:MMEDate.date name:eventName attributes:attributes];
+#pragma mark - Generic Events
+
++ (instancetype)eventWithAttributes:(NSDictionary *)attributes {
+    NSError *eventError = nil;
+    MMEEvent *newEvent = [MMEEvent eventWithAttributes:attributes error:&eventError];
+    if (eventError != nil) {
+        [MMEEventsManager.sharedManager reportError:eventError];
+    }
+
+    return newEvent;
+}
+
++ (instancetype)eventWithAttributes:(NSDictionary *)attributes error:(NSError **)error {
+    return [MMEEvent.alloc initWithAttributes:attributes error:error];
 }
 
+#pragma mark - Custom Events
+
 + (instancetype)turnstileEventWithAttributes:(NSDictionary *)attributes {
-    MMEEvent *turnstileEvent = [[MMEEvent alloc] init];
-    turnstileEvent.name = MMEEventTypeAppUserTurnstile;
-    turnstileEvent.attributes = attributes;
-    return turnstileEvent;
+    NSMutableDictionary *eventAttributes = attributes.mutableCopy;
+    eventAttributes[MMEEventKeyEvent] = MMEEventTypeAppUserTurnstile;
+
+    return [MMEEvent eventWithAttributes:eventAttributes];
 }
 
-+ (instancetype)telemetryMetricsEventWithDateString:(NSString *)dateString attributes:(NSDictionary *)attributes {
-    MMEEvent *telemetryMetrics = [[MMEEvent alloc] init];
-    telemetryMetrics.name = MMEEventTypeTelemetryMetrics;
-    NSMutableDictionary *commonAttributes = [NSMutableDictionary dictionary];
-    commonAttributes[MMEEventKeyEvent] = telemetryMetrics.name;
-    commonAttributes[MMEEventKeyCreated] = dateString;
-    [commonAttributes addEntriesFromDictionary:attributes];
-    telemetryMetrics.attributes = commonAttributes;
-    return telemetryMetrics;
++ (instancetype)visitEventWithAttributes:(NSDictionary *)attributes {
+    NSMutableDictionary *eventAttributes = attributes.mutableCopy;
+    eventAttributes[MMEEventKeyEvent] = MMEEventTypeVisit;
+
+    return [MMEEvent eventWithAttributes:eventAttributes];
 }
 
-+ (instancetype)locationEventWithAttributes:(NSDictionary *)attributes instanceIdentifer:(NSString *)instanceIdentifer commonEventData:(MMECommonEventData *)commonEventData {
+#pragma mark - Crash Events
+
++ (instancetype)crashEventReporting:(NSError *)eventsError error:(NSError **)createError {
+    // start with common event data
+    NSMutableDictionary *crashAttributes = MMECommonEventData.commonEventData;
+    crashAttributes[MMEEventKeyEvent] = MMEEventMobileCrash;
```

Comment:
Correct, the exceptions are wrapped up into `NSError` objects, and we generate a report from that, so that networking, filesystem or encode/decode errors can be collected as well.

`PLCrashReporter` for e.g. could be added to catch crashes, then this method used to send back a report on app launch if a crash file is left behind.